eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -679,3 +631,145 @@ fromPrevGovActionIds (PrevGovActionIds pforest@(PForest _ _ _ _)) =
         & pfHardForkL . prRootL .~ pfHardFork
         & pfCommitteeL . prRootL .~ pfCommittee
         & pfConstitutionL . prRootL .~ pfConstitution
+
+---------------------
+-- Debugging tools --
+---------------------
+
+-- | Wrapper type, which serves as a composition of @`Tree` . `StrictMaybe`@
+--
+-- Also its Show instance will print a nice tree structure.
+data TreeMaybe a = TreeMaybe {unTreeMaybe :: Tree (StrictMaybe a)}
+  deriving (Eq)
+
+instance Show (TreeMaybe (GovPurposeId p era)) where
+  show = ("\n" <>) . drawTree . fmap showGovPurposeId . unTreeMaybe
+    where
+      showGovPurposeId = \case
+        SNothing -> "x"
+        SJust (GovPurposeId govActionId) -> T.unpack (govActionIdToText govActionId)
+
+-- | Partial version of `toPForestEither`
+toPForest :: (Era era, HasCallStack) => Proposals era -> PForest TreeMaybe era
+toPForest = either error id . toPForestEither
+
+-- | Convert `Proposals` into a valid `Tree`
+toPForestEither :: Era era => Proposals era -> Either String (PForest TreeMaybe era)
+toPForestEither Proposals {pProps, pRoots, pGraph} = do
+  unless (OMap.invariantHolds' pProps) $ Left "OMap invariant is violated"
+
+  (pfPParamUpdate, nodesPParamUpdate) <-
+    toPTree (pfPParamUpdate pRoots) (unPGraph (pfPParamUpdate pGraph))
+  (pfHardFork, nodesHardFork) <-
+    toPTree (pfHardFork pRoots) (unPGraph (pfHardFork pGraph))
+  (pfCommittee, nodesCommittee) <-
+    toPTree (pfCommittee pRoots) (unPGraph (pfCommittee pGraph))
+  (pfConstitution, nodesConstitution) <-
+    toPTree (pfConstitution pRoots) (unPGraph (pfConstitution pGraph))
+
+  let allNodes =
+        Set.unions
+          [ Set.map unGovPurposeId nodesPParamUpdate
+          , Set.map unGovPurposeId nodesHardFork
+          , Set.map unGovPurposeId nodesCommittee
+          , Set.map unGovPurposeId nodesConstitution
+          ]
+      propsMap = OMap.toMap pProps
+      guardUnknown = do
+        let unknown = allNodes Set.\\ Map.keysSet propsMap
+        unless (null unknown) $ do
+          Left $ "Discovered unrecognized nodes: " ++ show unknown
+      guardUnique = do
+        let sumSizes =
+              sum
+                [ Set.size nodesPParamUpdate
+                , Set.size nodesHardFork
+                , Set.size nodesCommittee
+                , Set.size nodesConstitution
+                ]
+        unless (Set.size allNodes == sumSizes) $ do
+          Left $
+            "Duplicate govActionIds found between different purposes: "
+              ++ show (sumSizes - Set.size allNodes)
+  guardUnknown
+  guardUnique
+  pure PForest {pfPParamUpdate, pfHardFork, pfCommittee, pfConstitution}
+
+toPTree :: (Ord a, Show a) => PRoot a -> Map a (PEdges a) -> Either String (TreeMaybe a, Set a)
```

Comment:
Sorry, I am not sure what you mean. What would that property test be for?
`toPTree` is the thing that checks the invariant. What would this unfolder be checking?

Reply:
Oh nevermind, I was thinking from the perspective of constructing the tree starting from the root, but by using the library function just to test its equivalence with our implementation. It was just an idea, and I wasn't too sure.

Reply:
Current implementation is essentially `unfoldTreeM`, except the loop is written manually.

We don't need to test that this conversion happens correctly. It would just be duplicating the logic.